### PR TITLE
flamenco: back input regions and bincode valloc with spad

### PR DESCRIPTION
**Pull request by yufeng-jump**

### CODE REVIEW

File: src/flamenco/runtime/fd_runtime_spad.c [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#include "fd_runtime_spad.h"
+
```
Comment (ibhatt-jumptrading) [inline src/flamenco/runtime/fd_runtime_spad.c:2]: this is nice

File: src/flamenco/runtime/program/fd_bpf_loader_program.c [hunk 2]
```diff
@@ -236,11 +237,13 @@ calculate_heap_cost( ulong heap_size, ulong heap_cost, int * err ) {
 int
 fd_deploy_program( fd_exec_instr_ctx_t * instr_ctx,
                    uchar const *         programdata,
-                   ulong                 programdata_size ) {
+                   ulong                 programdata_size,
+                   fd_valloc_t           valloc /* Majority of use cases this is a txn ctx spad under the hood; for native to bpf migration this is scratch backed since it's an epoch level event. */
```
Comment (ibhatt-jumptrading) [inline src/flamenco/runtime/program/fd_bpf_loader_program.c:241]: We also need to bound out the epoch level events too. Probably makes sense to do in a follow up PR

File: src/flamenco/runtime/context/fd_exec_instr_ctx.h [hunk 3]
```diff
@@ -25,7 +25,7 @@ struct __attribute__((aligned(8UL))) fd_exec_instr_ctx {
 
   fd_funk_txn_t * funk_txn;
   fd_acc_mgr_t *  acc_mgr;
-  fd_valloc_t     valloc;
+  fd_valloc_t     valloc; /* Scratch-backed valloc TODO: migrate to transaction spad */
```
Comment (ibhatt-jumptrading) [inline src/flamenco/runtime/context/fd_exec_instr_ctx.h:28]: Can we just set the instr ctx valloc to the spad valloc

File: src/flamenco/runtime/fd_runtime_spad.c (hunk 1 above)
Comment (ibhatt-jumptrading) [reply src/flamenco/runtime/fd_runtime_spad.c:2]: maybe we add this to util and just generally expose a valloc for spad? There's nothing necessarily runtime specific about this

File: src/flamenco/runtime/context/fd_exec_txn_ctx.h [hunk 4]
```diff
@@ -64,7 +64,8 @@ struct __attribute__((aligned(8UL))) fd_exec_txn_ctx {
 
   fd_funk_txn_t *       funk_txn;
   fd_acc_mgr_t *        acc_mgr;
-  fd_valloc_t           valloc;
+  fd_valloc_t           valloc;                                      /* Workspace-backed valloc TODO: migrate to spad */
```
Comment (ibhatt-jumptrading) [inline src/flamenco/runtime/context/fd_exec_txn_ctx.h:67]: Again can this always just be equal to the spad valloc

File: src/flamenco/runtime/tests/fd_exec_sol_compat.c [hunk 5]
```diff
@@ -65,7 +65,12 @@ sol_compat_wksp_init( void ) {
   wksp = fd_wksp_new_anonymous( FD_SHMEM_NORMAL_PAGE_SZ, 65536UL * 8UL, fd_shmem_cpu_idx( fd_shmem_numa_idx( cpu_idx ) ), "wksp", 0UL );
   assert( wksp );
 
-  spad_mem = fd_wksp_alloc_laddr( wksp, FD_SPAD_ALIGN, fd_spad_footprint( MAX_TX_ACCOUNT_LOCKS * fd_ulong_align_up( FD_ACC_TOT_SZ_MAX, FD_ACCOUNT_REC_ALIGN ) ), 3 ); /* 1342191744 B */
+  // FD_LOG_WARNING(( "footprint min %lu max %lu default %lu fuzz %lu",
```
Comment (ibhatt-jumptrading) [inline src/flamenco/runtime/tests/fd_exec_sol_compat.c:68]: remove log?

File: src/flamenco/runtime/tests/fd_exec_instr_test.c [hunk 6]
```diff
@@ -1657,6 +1670,7 @@ fd_exec_vm_syscall_test_run( fd_exec_instr_test_runner_t * runner,
   if( !fd_exec_test_instr_context_create( runner, ctx, input_instr_ctx, alloc, skip_extra_checks ) )
     goto error;
```
Comment (ibhatt-jumptrading) [inline src/flamenco/runtime/tests/fd_exec_instr_test.c:1671]: I'm pretty sure we can just use spad here too (like in the valloc)

File: src/flamenco/runtime/fd_runtime.c [hunk 7]
```diff
@@ -2120,18 +2121,27 @@ fd_runtime_execute_txns_in_waves_tpool( fd_exec_slot_ctx_t * slot_ctx,
...
 
         /* Assign out spads to the transaction contexts */
         for( ulong i=0UL; i<wave_task_infos_cnt; i++ ) {
           wave_task_infos[i].spads = spads;
         }
 
+        // Dump txns in waves
+        if( dump_txn ) {
+          for( ulong i = 0; i < wave_task_infos_cnt; ++i ) {
+            // TODO this dumper still relies on scratch, move to spad?
```
Comment (ibhatt-jumptrading) [inline src/flamenco/runtime/fd_runtime.c:2139]: eh we can leave this in scratch, its not in the critical path

File: src/flamenco/runtime/fd_runtime.h [hunk 8]
```diff
@@ -95,6 +97,170 @@ FD_STATIC_ASSERT( FD_ACCOUNT_REC_ALIGN>=FD_ACCOUNT_REC_DATA_ALIGN, account_rec_d
 FD_STATIC_ASSERT( (offsetof(fd_account_rec_t, meta)%FD_ACCOUNT_META_ALIGN)==0,     account_rec_meta_offset );
 FD_STATIC_ASSERT( (offsetof(fd_account_rec_t, data)%FD_ACCOUNT_REC_DATA_ALIGN)==0, account_rec_data_offset );
 
+#define MAX_PERMITTED_DATA_INCREASE (10240UL) // 10KB
```
Comment (ibhatt-jumptrading) [inline src/flamenco/runtime/fd_runtime.h:100]: 🐐

File: src/flamenco/runtime/program/fd_bpf_loader_serialization.h [hunk 9]
```diff
@@ -3,9 +3,8 @@
 
 #include "../../fd_flamenco_base.h"
 #include "../../vm/fd_vm.h"
+#include "../fd_runtime.h"
```
Comment (ibhatt-jumptrading) [inline src/flamenco/runtime/program/fd_bpf_loader_serialization.h:6]: needed?

**Review by ibhatt-jumptrading — COMMENTED** [review]
I think mostly lgtm, I think it would be a lot better if we just stored a handle to the spad for the instr_ctx and txn_ctx instead of having to handle and pass around an spad pointer and a scratch valloc

File: src/flamenco/runtime/program/fd_bpf_loader_program.c (hunk 2 above)
Comment (yufeng-jump) [reply src/flamenco/runtime/program/fd_bpf_loader_program.c:241]: Agreed.

File: src/flamenco/runtime/program/fd_bpf_loader_serialization.h (hunk 9 above)
Comment (yufeng-jump) [reply src/flamenco/runtime/program/fd_bpf_loader_serialization.h:6]: good catch.  just need it in serialization.c and not the .h

File: src/flamenco/runtime/tests/fd_exec_instr_test.c (hunk 6 above)
Comment (yufeng-jump) [reply src/flamenco/runtime/tests/fd_exec_instr_test.c:1671]: Let's migrate these non critical path test harness allocations in a future PR, and keep this PR about live or replay transaction execution?

File: src/flamenco/runtime/context/fd_exec_instr_ctx.h (hunk 3 above)
Comment (yufeng-jump) [reply src/flamenco/runtime/context/fd_exec_instr_ctx.h:28]: Some instruction level test harnesses don't give us spads and instead give us libc malloc valloc and etc.  We can do that change in a future PR?  I'll update the comment to reflect that.

Comment (yufeng-jump) [reply src/flamenco/runtime/context/fd_exec_instr_ctx.h:28]: Okay I renamed this to test_only_valloc and left a comment to make it clear.

**Review by ibhatt-jumptrading — APPROVED** [review]

**Pull request merged**